### ISSUE TITLE
Expose generic type system for plugins

### ISSUE THREAD

**Issue by alvar-osz** [graphql-hive/envelop#106]
**Is your feature request related to a problem? Please describe.**

Currently, when creating the envelop with typescript there are some plugins that have no type information other than unknown, but they do return a `Plugin<DefaultContext>` meaning they do have some extensibility potential.

By pain point is using `useExtendedSchema` which has a function signature of `(unknown) => unknown` so when I pass my typed function, it says it's not compatible.

**Describe the solution you'd like**
Expose type generics to extend the plugin and that they match the context type coming from envelop.

```typescript
envelop<Context extends DefaultContext = DefaultContext>(options: {
  plugins: Plugin<Context>[];
}) => Envelop<Context>;

useSchema<Context extends DefaultContext = DefaultContext>(schema: GraphQLSchema) => Plugin<Context>;

useExtendSchema<Context extends DefaultContext = DefaultContext, PrevContext  extends DefaultContext = DefaultContext>(createContext: (PrevContext) => Context) => Plugin<Context>;

// It could even go like this, since the previous context is something the entire envelop will receive at some point

envelop<
  Context extends DefaultContext = DefaultContext,
  PrevContext  extends DefaultContext = DefaultContext
>(options: {
  plugins: Plugin<Context, PrevContext>[];
}) => Envelop<Context, PrevContext>;

useSchema<Context extends DefaultContext = DefaultContext, PrevContext  extends DefaultContext = DefaultContext>(schema: GraphQLSchema) => Plugin<Context, PrevContext>;

useExtendSchema<Context extends DefaultContext = DefaultContext, PrevContext  extends DefaultContext = DefaultContext>(createContext: (PrevContext) => PromiseLike<Context>) => Plugin<Context, PrevContext>;

// This could even help type 

const {
        parse,
        validate,
        contextFactory,
        execute,
        schema
      } = getEnveloped();

contextFactory // (PrevContext) => PromiseLike<Context>

```

**Describe alternatives you've considered**
Using `as` to cast the types but it's not a very nice solution.

**Comment by dotansimha** [comment]
Totally agree! I started with the basics and removed some of the unused, but you are definitely right that this could be improved.

**Comment by dotansimha** [comment]
I addressed this issue while doing some changes in https://github.com/dotansimha/envelop/pull/116 

Now each plugin can specify the ContextType relevant to it, and it's fully type-safe :)

**Comment by dotansimha** [comment]
Fixed in `@envelop/core@0.1.3`.

**Comment by alvar-osz** [comment]
Great thank you @dotansimha 😎 

Could you extend the exposed generic type to the current plugins, already located in `core` and `plugins`?

**Comment by dotansimha** [comment]
> Great thank you @dotansimha 😎
> 
> Could you extend the exposed generic type to the current plugins, already located in `core` and `plugins`?

I'm not sure that I fully understood, can you please provide an example? You can see the `PluginContext` in use here for an example: 
https://github.com/dotansimha/envelop/blob/main/packages/plugins/rate-limiter/src/index.ts#L22-L24

**Comment by alvar-osz** [comment]
Here for example, https://github.com/dotansimha/envelop/blob/5f27ea0439e3460e6d2b7da54b7995bdbd1f8761/packages/core/src/plugins/use-extend-context.ts#L5

That was the plugin that originally made me realize there was no exposed context type.

**Comment by dotansimha** [comment]
> Here for example,
> 
> https://github.com/dotansimha/envelop/blob/5f27ea0439e3460e6d2b7da54b7995bdbd1f8761/packages/core/src/plugins/use-extend-context.ts#L5
> 
> That was the plugin that originally made me realize there was no exposed context type.

Got it. I took care only at the part at the Plugin API, and not as a whole. Will definitely look into that as well. Thanks @ajotaos !

**Comment by dotansimha** [comment]
Available in `@envelop/core@0.4.0`
